### PR TITLE
fix node recipes, add run_list_expansion and cookbooks

### PR DESCRIPTION
**Pull request by lamont-granquist**
* fix node[:recipes] 
* add node[:cookbooks]
* add node[:expanded_run_list]

### CODE REVIEW

**Comment by jaymzh** [comment]
Didn't this make it into 12 already?

**Comment by lamont-granquist** [comment]
nope

**Comment by lamont-granquist** [comment]
@chef/client-core review?  the travis failure is the e2e tests failing hard...

**Comment by ranjib** [comment]
:+1:

File: lib/chef/node.rb [hunk 1]
```diff
@@ -362,7 +372,8 @@ def expand!(data_source = 'server')
 
       self.tags # make sure they're defined
 
-      automatic_attrs[:recipes] = expansion.recipes
+      automatic_attrs[:recipes] = expansion.recipes.with_fully_qualified_names_and_version_constraints
```
Comment (jaymzh) [inline lib/chef/node.rb]: isn't this blanketly overwriting what line 265 does?

**Comment by jaymzh** [comment]
Out of curiosity, why did you think this would improve perf? I don't see why this would affect it.

File: lib/chef/node.rb (hunk 1 above)
Comment (lamont-granquist) [inline lib/chef/node.rb]: no, this is where its initialized, then 265 is applied on top of that.

**Comment by lamont-granquist** [comment]
It doesn't affect perf, that was an idea that serdar had that line 265 was thrashing the node object somehow.  I backed out the different approach to that line because it wasn't useful for perf at all.

**Comment by lamont-granquist** [comment]
@chef/client-core think this still needs one more +1 assuming the travis failures go away...

**Comment by btm** [comment]
##### fix node[:recipes]

Is the fix that we're making default recipes in node[:recipes] fully qualified?

##### add node[:cookbooks]

Is this early desired versus current node state work?

##### add node[:expanded_run_list]

It looks like the difference between node[:recipes] and node[:expanded_run_list] is that node[:recipes] includes "included recipes" but node[:expanded_run_list] does not. We added that functionality in https://github.com/chef/chef/pull/1128. Makes sense, but is there a bug/issue where we've made the case this is necessary?

**Comment by lamont-granquist** [comment]
yeah there's an edge case in node[:recipes] where you can wind up with both 'foo' and 'foo::default' because it gets initialized with the expanded_run_list and later gets included with the fully qualified recipe name.  this fixes that so its all fully qualified.  there's bugs on this in the old ticket tracker.

node[:cookbooks] is very commonly requested so that people can easily audit which cookbooks are in use or not via querying the nodes.  also old tickets in the ticket tracker, probably issues here, threads on the mailing list, etc.

node[:expanded_run_list] is not really driven by customer requests AFAIK, i think dan suggested that one.

**Comment by danielsdeleo** [comment]
The expanded run list is a thing we de facto used to have, when we set the 'cookbooks' attribute right after expansion. This means you could do limited lookahead on recipes that weren't yet loaded but were going to be (limited because you don't know if a "future" recipe will `include_recipe` something else). Is anyone using that? I don't know, but it's a thing that's not hard to provide so I think it's best we don't remove it unless there is some additional argument for doing so.

**Comment by lamont-granquist** [comment]
It was actually getting loaded into node[:recipes] and we broke that in 12.0.0 and i haven't seen a bug report asking for it back...

**Comment by danielsdeleo** [comment]
I could buy that, though Chef 12+ adoption hasn't been that great so far, so I'm not sure if that's correlation or causation.

**Comment by lamont-granquist** [comment]
Yeah, I can go either way, thing is that node[:expanded_run_list] won't hurt much and might help the one person who really wants to see it.  OTOH, there's no customer request so no driving requirement so YAGNI...  But its so tiny I'm not worried about it either way...  I tend to think that bias towards better instrumentation wins slightly.

**Comment by ranjib** [comment]
@danielsdeleo i dont think anyone is using that look ahead mechanism. from what i recall there are couple of discussions on this in chef user mailing list and we clarified include_recipe effect on node.recipes. But still we can call it out as part of the release notes

**Comment by lamont-granquist** [comment]
Ah, right, so a point of clarification is that I solved the node[:recipes] problem by making the initial initialization with the expanded_run_list be fully qualified.  So I didn't fully lazy node[:recipes] to only happen at include_recipe time.  Then because include_recipe already is protected against adding duplicate recipes to node[:recipes] and they both agree on canonicalization now it should eliminate the dups.  If you look at the code, node[:recipes] and node[:expanded_run_list] are identical when they get initialized, and then later node[:recipes] collects recipes via include_recipe...

**Comment by lamont-granquist** [comment]
so we all :+1: or no?

**Comment by danielsdeleo** [comment]
Do it :+1:

**Comment by mihasya** [comment]
FYI this breaks compatibility for certain use cases of Chef Search.

Nodes that had previously had a runlist of just `recipe[cookbookname]` and could be looked up using a search for `recipes:cookbookname` will now report their runlist as recipe[cookbookname::default]` causing the search to fail. Adding a `*` to the end of the search does fix it.

Semi-related: I still haven't figured out how to explicitly search for `cookbookname::default` as the `::` messes with the syntax parser. Escaping them fixes it when invoking from `knife`, but we haven't found a way to make it work in a recipe or `chef-shell`. 

This went into Chef 12.4, and there appears to be absolutely zero mention of it in [the release notes](https://github.com/chef/chef/pull/1401). The `CHANGELOG` entry doesn't make it clear that this would be a side effect either, leading me to believe it's probably not intended.

We've fixed our lookup helper to work with either version, but this may be worth another look..

**Comment by eherot** [comment]
Re-iterating what @mihasya said here: This broke compatibility for our operation. Recipes which had been configured to search for `recipe:cookbook_name` suddenly started coming up empty.

**Pull request merged**